### ISSUE TITLE
MCMC STAN

### ISSUE THREAD

**Issue by Craig44** [NIWAFisheriesModelling/CASAL2#104]
Just make a formal reminder, see if we can include STAN as a thirdparty library for use in MCMC. Reference to site
http://mc-stan.org/development/

**Comment by Craig44** [comment]
http://discourse.mc-stan.org/t/using-stan-with-stand-alone-models/3504/5

I have come across this discourse which would be helpful if we want to go down this route.

**Comment by Craig44** [comment]
https://github.com/kaskr/tmbstan/tree/master/tmbstan

This R package pretty much ports into STAN, so I believe this to be the best starting point, if we do not bring anyone in from the STAN dev team

**Comment by Craig44** [comment]
If you can't plug in Stan Cole Monnahan has written the functions for ADMB in the following scripts, but this could be a fairly big job. just adding for extra info
https://github.com/admb-project/admb/blob/master/src/nh99/nuts.cpp
https://github.com/admb-project/admb/blob/master/src/nh99/hmc_functions.cpp
https://github.com/admb-project/admb/blob/master/src/nh99/hmc.cpp
https://github.com/admb-project/admb/blob/master/src/nh99/hybmcmc.cpp

**Comment by Craig44** [comment]
Just to keep everything self contained, Trying to integrate Stan into Casal2 is difficult. Stan's functionality are setup to calculate gradients/optimise/MCMC. These algorithms use reverse Automatic differentiation, each time the chain rule is applied a destructor for AD variable gets called. This is a problem in Casal2 because all variables in the system are assigned as AD variables (even data) so after a gradient evaluation is done all our input data and constants are deleted. Stan gets around this by compiling for each different model configuration. I attempted to do this [here](https://github.com/Craig44/CASAL2/tree/IntegrateStan) I also set up a little standalone application [here](https://github.com/Craig44/UsingStancpp). I am thinking from here on if we want this functionality we follow the post above and migrate code from stan and admb ourselves.
